core/state_update.go

```diff
@@ -51,38 +52,91 @@ func (d *StateDiff) Length() uint64 {
 }
 
 func (d *StateDiff) Hash() *felt.Felt {
-	prefixHash, err := new(felt.Felt).SetString("STARKNET_STATE_DIFF0")
-	if err != nil {
-		// shouldn't happen
-		panic(err)
-	}
+	var digest crypto.PoseidonDigest
+
+	digest.Update(new(felt.Felt).SetBytes([]byte("STARKNET_STATE_DIFF0")))
 
-	hashElems := []*felt.Felt{prefixHash}
+	// updated_contracts = deployedContracts + replacedClasses
+	// Digest: [number_of_updated_contracts, address_0, class_hash_0, address_1, class_hash_1, ...].
+	numOfUpdatedContracts := uint64(len(d.DeployedContracts) + len(d.ReplacedClasses))
+	digest.Update(new(felt.Felt).SetUint64(numOfUpdatedContracts))
 
-	// updated contracts
-	numOfUpdatedContracts := len(d.DeployedContracts) + len(d.ReplacedClasses)
-	hashElems = append(hashElems, new(felt.Felt).SetUint64(uint64(numOfUpdatedContracts)))
-	// todo sort
-	for addr, classHash := range d.DeployedContracts {
-		hashElems = append(hashElems, &addr, classHash)
+	// The sequencer guarantees that a contract cannot be:
+	// - deployed twice,
+	// - deployed and have its class replaced in the same state diff, or
+	// - have its class replaced multiple times in the same state diff.
+	updatedContracts := make(map[felt.Felt]*felt.Felt)
+	for k, v := range d.DeployedContracts {
+		updatedContracts[k] = v
 	}
-	// todo sort
-	for addr, classHash := range d.ReplacedClasses {
-		hashElems = append(hashElems, &addr, classHash)
+	for k, v := range d.ReplacedClasses {
+		updatedContracts[k] = v
+	}
+
+	sortedUpdatedContractsHashes := sortedFeltKeys(updatedContracts)
+	for _, hash := range sortedUpdatedContractsHashes {
+		digest.Update(&hash, updatedContracts[hash])
 	}
 
 	// declared classes
-	numOfDeclaredClasses := len(d.DeclaredV1Classes)
-	hashElems = append(hashElems, new(felt.Felt).SetUint64(uint64(numOfDeclaredClasses)))
-	for classHash, compiledClasshash := range d.DeclaredV1Classes {
-		hashElems = append(hashElems, &classHash, compiledClasshash)
+	// Digest: [number_of_declared_classes, class_hash_0, compiled_class_hash_0, class_hash_1, compiled_class_hash_1,
+	// ...].
+	numOfDeclaredClasses := uint64(len(d.DeclaredV1Classes))
+	digest.Update(new(felt.Felt).SetUint64(numOfDeclaredClasses))
+
+	sortedDeclaredV1ClassHashes := sortedFeltKeys(d.DeclaredV1Classes)
+	for _, classHash := range sortedDeclaredV1ClassHashes {
+		digest.Update(&classHash, d.DeclaredV1Classes[classHash])
+	}
+
+	// deprecated_declared_classes
+	// Digest: [number_of_old_declared_classes, class_hash_0, class_hash_1, ...].
+	numOfDeclaredV0Classes := uint64(len(d.DeclaredV0Classes))
+	digest.Update(new(felt.Felt).SetUint64(numOfDeclaredV0Classes))
+
+	// Todo: consider make a copy of d.DeclaredV0Classes?
+	slices.SortFunc(d.DeclaredV0Classes, func(a, b *felt.Felt) int { return a.Cmp(b) })
+	digest.Update(d.DeclaredV0Classes...)
+
+	// Placeholder values
+	digest.Update(new(felt.Felt).SetUint64(1), new(felt.Felt).SetUint64(0))
```

Comment:
Is this "number_of_DA_modes, DA_mode_0"?

Reply:
No idea, see here: https://github.com/starkware-libs/starknet-api/blob/15219992b54b170dcf80fa5cceab56847ea2f7f0/src/block_hash/state_diff_hash.rs#L34-L35